File: PHP/leetcode/pizza-with-3n-slices/solution.php
class Solution {

    /**
     * @param Integer[] $slices
     * @return Integer
     */
    function maxSizeSlices($slices) {
        $total = count($slices);
        $k = intdiv($total, 3);
        if ($k == 0) return 0;
        $a = array_slice($slices, 0, $total - 1);
        $b = array_slice($slices, 1);
        return max($this->best($a, $k), $this->best($b, $k));
    }

    private function best($nums, $k) {
        $n = count($nums);
        $NEG = -(1 << 60);
        $dp = array();
        for ($i = 0; $i <= $n; $i++) {
            $dp[$i] = array_fill(0, $k + 1, $NEG);
            $dp[$i][0] = 0;
        }
        for ($i = 1; $i <= $n; $i++) {
            for ($j = 1; $j <= $k; $j++) {
                $skip = $dp[$i - 1][$j];
                if ($i >= 2) {
                    $prev = $dp[$i - 2][$j - 1];
                } else {
                    $prev = ($j == 1) ? 0 : $NEG;
                }
                $take = $prev + $nums[$i - 1];
                $dp[$i][$j] = max($skip, $take);
            }
        }
        return $dp[$n][$k];
    }
}

File: Rust/leetcode/pizza-with-3n-slices/solution.rs
impl Solution {
    pub fn max_size_slices(slices: Vec<i32>) -> i32 {
        let total = slices.len();
        let k = total / 3;
        if k == 0 {
            return 0;
        }
        let a = slices[..total - 1].to_vec();
        let b = slices[1..].to_vec();
        pizza_best(&a, k).max(pizza_best(&b, k))
    }
}

fn pizza_best(nums: &[i32], k: usize) -> i32 {
    let n = nums.len();
    let neg: i64 = i64::MIN / 4;
    let mut dp = vec![vec![neg; k + 1]; n + 1];
    for i in 0..=n {
        dp[i][0] = 0;
    }
    for i in 1..=n {
        for j in 1..=k {
            let skip = dp[i - 1][j];
            let prev = if i >= 2 {
                dp[i - 2][j - 1]
            } else if j == 1 {
                0
            } else {
                neg
            };
            let take = prev + nums[i - 1] as i64;
            dp[i][j] = skip.max(take);
        }
    }
    dp[n][k] as i32
}
